### PR TITLE
[BREAKING] feat!: allow vaults to opt-out of new issues

### PR DESCRIPTION
**Pull request by sander2**
The breaking change here is the VaultStatus enum inside vault structs. The `Active` status is now `Active(bool)`, where `bool=true` indicates that the vault accepts new issues. This approach was taken of splitting it into two separate variants (e.g. `Active` and `Exiting`), because usually when we read the `status`, we don't need to differentiate between the two. The `get_active_vault` function that we use extensively would have to be turned into `get_active_or_exiting_vault` otherwise, which is just.. ugly. Not to mention that `exiting` is not 100% accurate.

Also note that unlike what Dom asked for, I did not include `BannedUntil` in the status enum. This is because of two reasons: first, after a temporary ban, you would be unable to return to the previous state, since we wouldn't remember whether or not we accepted new issues right before the ban. The second reason is that we would have no easy way to change the status after the ban is over. We could perhaps do it within the `get_active_vault` function, but then it'd only be updated after a dispatchable function is called on that vault. If the clients or UI read the struct directly, they would get a wrong status.

Finally, note (and please check) that I added checks to filter out vaults that do not accept new issues in the relevant rpc functions. I chose to do it here, rather than inside a deeper function like `get_issuable_tokens`, since otherwise also `accept_replace` and `auction_replace` wouldn't work, and I saw no reason to prevent the vaults from doing that.

### CODE REVIEW

File: crates/vault-registry/src/lib.rs [hunk 1]
```diff
@@ -156,29 +157,32 @@ decl_module! {
         fn on_runtime_upgrade() -> Weight {
             use frame_support::{migration::StorageKeyIterator, Blake2_128Concat};
 
-            if matches!(Self::storage_version(), Version::V0 | Version::V1) {
-                StorageKeyIterator::<T::AccountId, VaultV1<T::AccountId, T::BlockNumber, PolkaBTC<T>, DOT<T>>, Blake2_128Concat>::new(<Vaults<T>>::module_prefix(), b"Vaults")
+            if matches!(Self::storage_version(), Version::V2) {
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:160]: Note that we haven't yet run the V2 storage migration, so we should bundle this upgrade into V2.

File: crates/vault-registry/src/lib.rs [hunk 2]
```diff
@@ -288,6 +295,23 @@ decl_module! {
             Ok(())
         }
 
+        /// Configures whether or not the vaul accepts new issues.
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:298]:
```suggestion
        /// Configures whether or not the vault accepts new issues.
```

File: crates/vault-registry/src/lib.rs [hunk 3]
```diff
@@ -288,6 +295,23 @@ decl_module! {
...
+        /// * `accept_new_issues` - true indicates that the vault accepts new issues
+        ///
+        /// # Weight: `O(1)`
+        #[weight = <T as Config>::WeightInfo::accept_new_issues()]
+        #[transactional]
+        fn accept_new_issues(origin, accept_new_issues: bool) -> DispatchResult  {
+            let vault_id = ensure_signed(origin)?;
+            let mut vault = Self::get_active_rich_vault_from_id(&vault_id)?;
+            vault.set_accept_new_issues(accept_new_issues)?;
+            Ok(())
```
Comment (gregdhill) [inline crates/vault-registry/src/lib.rs:312]:
```suggestion
            vault.set_accept_new_issues(accept_new_issues)
 ```

**Comment by nud3l** [comment]
> The breaking change here is the VaultStatus enum inside vault structs. The `Active` status is now `Active(bool)`, where `bool=true` indicates that the vault accepts new issues. This approach was taken of splitting it into two separate variants (e.g. `Active` and `Exiting`), because usually when we read the `status`, we don't need to differentiate between the two. The `get_active_vault` function that we use extensively would have to be turned into `get_active_or_exiting_vault` otherwise, which is just.. ugly. Not to mention that `exiting` is not 100% accurate.
> 
> 
> 
> Also note that unlike what Dom asked for, I did not include `BannedUntil` in the status enum. This is because of two reasons: first, after a temporary ban, you would be unable to return to the previous state, since we wouldn't remember whether or not we accepted new issues right before the ban. The second reason is that we would have no easy way to change the status after the ban is over. We could perhaps do it within the `get_active_vault` function, but then it'd only be updated after a dispatchable function is called on that vault. If the clients or UI read the struct directly, they would get a wrong status.
> 
> 

Ok, this makes sense to me. So in the stats and lib we have to stitch together the state from the current parachain height and the banned_until item. But that is cheaper than the on-chain operation, so that's fine. We need a ticket for this in the lib/stats though. Can you please add one?


> 
> Finally, note (and please check) that I added checks to filter out vaults that do not accept new issues in the relevant rpc functions. I chose to do it here, rather than inside a deeper function like `get_issuable_tokens`, since otherwise also `accept_replace` and `auction_replace` wouldn't work, and I saw no reason to prevent the vaults from doing that.

LGTM

**Review by nud3l — APPROVED** [review]

**Pull request merged**